### ISSUE TITLE
Addition of MobileNets to Keras Applications

### ISSUE THREAD

**Issue by titu1994** [fchollet/keras#6968]
I've recently finished implementing MobileNets for Keras, and was wondering if it could be added to Keras proper or the Deep Learning Models repository. 

Weights for ImageNet are available, and ported over (from the Tensorflow checkpoints) from the tf-slim  repository. 

There is an issues with adding it to Keras. Namely : 
1) Use of custom DepthwiseConvolution2D layer (1st part of SeparableConvolution2D). While I have tested the Keras model on random images, and ensured it works, we would have to add another layer to Keras, and the accompanying tests.
2) **EDIT** The model has 1001 output classes, but `decode_predictions` accepts only 1000 classes. I got around this limitation by selecting just the first 1000 predictions in `predict_imagenet.py`

@fchollet I've considered adding this to contrib, but the main benefit of having this model in Keras proper is its efficiency. I found it to be far faster and easier to fine tune as compared to InceptionV3 or ResNet50, which can't even be loaded onto smaller GPUs (such as mine with only 4 GB of memory), with only a minimal loss in performance. 

Other benefits are obvious as to deployment on mobile devices (with the Keras integration into tensorflow 1.2 being almost complete in the latest RC, I imagine it would be trivial to train on Keras and export to Tensorflow and then to Android / iOS (hopefully)).

Relevant Paper - [MobileNets: Efficient Convolutional Neural Networks for Mobile Vision Applications](https://arxiv.org/abs/1704.04861)
Repository - https://github.com/titu1994/MobileNetworks
Weights were ported from repository - https://github.com/Zehaos/MobileNet

**Comment by titu1994** [comment]
@fchollet I have finished porting all 16 weights for the MobileNets that were posted yesterday to Keras. All 16 work out of the box, but the only drawback is that I could not get the weights to work with `channels_first` data format.

Should I send a PR to Keras or to Contrib?

**Comment by ahundt** [comment]
I'm curious what did you use to convert this?

**Comment by titu1994** [comment]
I modified a tensorflow function which originally simply prints out the numpy arrays in the checkpoints file to write all of the numpy arrays into a directory.

I then wrote a few scripts for convenience, such as removing the extra files like RMSProp or global_step or ExponentialMovingAverages.

This leads to 137 numpy arrays for each of the models. This I then wrote a script to correctly load the numpy arrays into the Keras model weights, and perform a sanity check to see if all possible weights were loaded or not.

After that, a final script creates a version of the weight with the last few layers removed to create no_top weights.

The main hassle was noticing that tensorflow saves the weights of BatchNorm as `[beta, gamma, running_mean, running_variance]`, whereas Keras expects them in the format `[gamme, beta, running_mean, running_variance]`. Since beta and gamma are both same shape, they pass the sanity check, but the predictions will be consistent garbage unless you correct it.
